[refman] Document the vernac `Type` command (#16681)
coq/coq: doc/sphinx/proof-engine/vernacular-commands.rst

.. cmd:: Check @term

   Displays the type of :n:`@term`. When called in proof mode, the term is

>    checked in the local context of the selected goal. This command will try
>    resolve an existential variables.

Reviewer: ```suggestion
   checked in the local context of the selected goal. This command tries
   to resolve existential variables.
```

Is there a short and accurate description of how Evars get resolved by `Check`?  Could you perhaps give a short example?  (so I can understand and perhaps to include in the doc.)

Author: ```coq
Class SomeType := {x : nat}.
#[export] Instance default_term : SomeType := {x := 0}.
Axiom some_predicate : SomeType -> Prop.
Axiom some_proof : forall (t : SomeType), some_predicate t.

Check (some_proof _).
(* some_proof default_term
     : some_predicate default_term *)

Type (some_proof _).
(* some_proof default_term
     : some_predicate default_term *)
```

It appears that `Type` does in fact resolve evars so my documentation of it is incorrect as `@SkySkimmer` says. The difference is this:

```coq
Class SomeType := {x : nat}.
(* #[export] Instance default_term : SomeType := {x := 0}. *)
Axiom some_predicate : SomeType -> Prop.
Axiom some_proof : forall (t : SomeType), some_predicate t.

Check (some_proof _).
(* some_proof ?t
     : some_predicate ?t
where
?t : [ |- SomeType] *)

Type (some_proof _).
(* Cannot infer this placeholder of type
"SomeType" (no type class instance
found). *)

```

Both resolve evars, however `Check` will not fail when evars are left, whereas `Type` will.

.. cmd:: Type @term

>    Like :cmd:`Check`, displays the type of :n:`@term` however without attempting
>    to resolve existential variables.

Reviewer: ```suggestion
   Displays the type of :n:`@term`, similar to :cmd:`Check`, but won't
   resolve existential variables.
```

>    to resolve existential variables.

Reviewer 2: Why do you think that it doesn't resolve existentials?
~~~coq
Class A := {}.
Local Declare Instance Ainst : A.

Type _ : A. (* Ainst : A *)
~~~

Type is like Check, but always uses the global env (Check uses the first focused goal env when there is one) and errors if there are unresolvable existentials.

Author: What about typeclass search? I seem to recall `Type` not going into typeclass search. Maybe I need to dig in the sources to see what is going on.

Reviewer 2: The example I posted is typeclass search

Reviewer 3: > Why do you think that it doesn't resolve existentials?

Probably because you wrote "you can use Type instead of Check to forbid evars" when you posted the [comment](https://coq.zulipchat.com/#narrow/stream/237977-Coq-users/topic/Declaring.20record.20constructor.20as.20Canonical/near/277299748) that prompted Ali to create #15881.

Author: @SkySkimmer is correct however. I've given an example displaying the difference in the reply to jfehrle above.

.. cmd:: Check @term

   Displays the type of :n:`@term`. When called in proof mode, the term is

>    checked in the local context of the selected goal. This command tries to

Reviewer 3: Probably worth making explicit that single numbered goal selectors can be used to select a goal.

   resolve existential variables as much as possible. When resolved, the term

>    will be checked again.

Reviewer 3: I don't know what to do with this last sentence. As is, it is probably insufficiently clear what that means to the reader and why it is so. So, without a better suggestion, I would suggest dropping it.

.. cmd:: Type @term

>    Displays the type of :n:`@term`, same as :cmd:`Check`, but will fail if all

Reviewer 3: Do you really mean "all" here or should it be "any"?

   existential variables are unable to be resolved. When resolved, the term will

>    be checked again.

Reviewer 3: Same here.
